src/collapse/dtrace.rs

```diff
@@ -113,19 +120,129 @@ impl Collapse for Folder {
     }
 }
 
-impl From<Options> for Folder {
-    fn from(opt: Options) -> Self {
-        Self {
-            cache_inlines: Vec::new(),
-            occurrences: FnvHashMap::with_capacity_and_hasher(512, fnv::FnvBuildHasher::default()),
-            stack: VecDeque::default(),
-            stack_str_size: 0,
-            opt,
+impl Folder {
+    fn collapse_single_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        let mut line = String::with_capacity(CAPACITY_LINE_BUFFER);
+
+        // skip header lines -- first empty line marks start of data
+        loop {
+            line.clear();
+            if reader.read_line(&mut line)? == 0 {
+                // We reached the end :( this should not happen.
+                warn!("File ended while skipping headers");
+                return Ok(());
+            };
+            if line.trim().is_empty() {
+                break;
+            }
+        }
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        Ok(())
+    }
+
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        debug_assert!(self.occurrences.is_concurrent());
+        debug_assert!(self.opt.nthreads > 1);
```

Comment:
Any particular reason you don't want these to be `assert!`?

Reply:
Yes - Since these are really "messages" for us (the library authors) that say - "Hey, don't call this private function in any code path where these conditions aren't met," I figured users don't need the checks in their release builds. There's nothing they (the users) can do to trigger or not trigger these asserts. If they ever get triggered it's because we made a mistake, which will be caught by testing in debug mode.

Reply:
I guess my thinking here is: they are not on a performance critical path, and without them users may get wrong results without knowing that something went wrong. If it crashes, at least they'll know, and have a way to report to us that something crashed (including where it crashed).